api/src/main/scala/Utils.scala

```diff
@@ -205,4 +207,22 @@ trait Utils {
       .query[SpatialKey, Tile, TileLayerMetadata[SpatialKey]](layerId)
       .where(Intersects(shape))
       .result
+
+  class MinWithoutNoData extends DoubleBinaryOperator {
+    override def applyAsDouble(left: Double, right: Double): Double =
+      if (isNoData(left))
+        if (isNoData(right)) Double.MaxValue
+        else right
+      else if (isNoData(right)) left
+      else math.min(left, right)
+  }
```

Comment:
All these operations seem to be working well and I don't think we should rewrite them, but when I looked up `isNoData` to check what it did I encountered some built-in summary ops in GeoTrellis:

https://github.com/locationtech/geotrellis/tree/master/raster/src/main/scala/geotrellis/raster/summary

Didn't delve too far into figuring out how to use them but the names, at least, suggest that they may be able to do similar operations. That said, this is working as expected so unless there are major performance gains to be had from using the built-in GeoTrellis versions I don't think we should rewrite. May be worth asking in #geotrellis

Reply:
I paired with @jamesmcclain for a bit this afternoon and this is what we came up with:

```scala
private def rasterSummary2(
  rasterLayers: Seq[TileLayerCollection[SpatialKey]],
  multiPolygon: MultiPolygon,
  opts: Rasterizer.Options
): Seq[Map[String, Double]] = {

  val p = multiPolygon.polygons.head

  val mins = rasterLayers.map { rl =>
    MinDoubleSummary.combineResults(rl.map { case (key, tile) =>
      val extent = rl.metadata.mapTransform(key)
      val rt = Raster(tile, extent)
      MinDoubleSummary.handlePartialTile(rt, p)
    })
  }
  val means = rasterLayers.map { rl =>
    MeanSummary.combineResults(rl.map { case (key, tile) =>
        val extent = rl.metadata.mapTransform(key)
        val rt = Raster(tile, extent)
        MeanSummary.handlePartialTile(rt, p)
    }).mean
  }
  val maxs = rasterLayers.map { rl =>
    MaxDoubleSummary.combineResults(rl.map { case (key, tile) =>
      val extent = rl.metadata.mapTransform(key)
      val rt = Raster(tile, extent)
      MaxDoubleSummary.handlePartialTile(rt, p)
    })
  }

  (mins zip means zip maxs) map { case ((min, avg), max) =>
    Map(
      "min" -> min,
      "avg" -> avg,
      "max" -> max
    )
  }
}
```

Here's a patch file of the entire change: [rasterPolygonalSummary.patch.txt](https://github.com/WikiWatershed/mmw-geoprocessing/files/1598410/rasterPolygonalSummary.patch.txt)

We found that the performance was slightly slower (but probably within noise range):

```bash
$ kj benchmark --rs
Timing HUC8 RasterSummary ->

HUC8 RasterSummary, run 1 -> 6.619486 s
HUC8 RasterSummary, run 2 -> 6.111745 s
HUC8 RasterSummary, run 3 -> 5.645586 s
HUC8 RasterSummary, run 4 -> 5.995264 s
HUC8 RasterSummary, run 5 -> 5.654279 s

HUC8 RasterSummary average -> 6.005272 s
```

I think we should stick with the code already in this PR for the following reasons:

* MultiPolygon handling: Since the Polygonal Summary only works for Polygons, we'll have to map over each polygon in the input, an additional loop (and potential slowdown)
* Rasterizer Options: The built-in methods do not take any Rasterizer Options, so we cannot treat raster pixels as areas, which may be required for some of our layers (the climatology ones are particularly low-res)
* Triple looping: Although this may be addressed by the compiler during optimization, at face value using the built-in methods is slower because we have to map over each layer for every aggregate value, whereas in our original implementation we're only mapping over every layer once.
* Pattern break: The current code follows the same pattern as all other operations in the geoprocessing service, whereas this would add a new pattern for little overall gain.

But this was still a very educational exercise and I feel more confident in our code thanks to it.

Reply:
:1st_place_medal:

Reply:
:+1: